internal/connector/connector.go

```diff
@@ -98,123 +101,79 @@ func (j *jwkCache) getJWK() (*jose.JSONWebKey, error) {
-
-			return
-		}
-
-		key, err := getJWK()
-		if err != nil {
-			logging.L.Debug("could not get jwk")
-			c.AbortWithStatus(http.StatusUnauthorized)
-
-			return
-		}
-
-		var claims struct {
-			jwt.Claims
-			claims.Custom
-		}
-
-		out := make(map[string]interface{})
-		if err := tok.Claims(key, &claims, &out); err != nil {
-			logging.L.Debug("invalid token claims")
-			c.AbortWithStatus(http.StatusUnauthorized)
-
-			return
-		}
-
-		err = claims.Claims.Validate(jwt.Expected{
-			Time: time.Now().UTC(),
-		})
+var JWKCacheRefresh = 5 * time.Minute
 
-		switch {
-		case errors.Is(err, jwt.ErrExpired):
-			logging.S.Debugf("expired JWT %s", err.Error())
-			c.AbortWithStatus(http.StatusUnauthorized)
+func (j *authenticator) Authenticate(req *http.Request) (claims.Custom, error) {
+	c := claims.Custom{}
+	authHeader := req.Header.Get("Authorization")
 
-			return
-		case err != nil:
-			logging.S.Debugf("invalid JWT %s", err.Error())
-			c.AbortWithStatus(http.StatusUnauthorized)
+	raw := strings.TrimPrefix(authHeader, "Bearer ")
+	if raw == "" {
+		return c, fmt.Errorf("no bearer token found")
+	}
 
-			return
-		}
+	tok, err := jwt.ParseSigned(raw)
+	if err != nil {
+		return c, fmt.Errorf("invalid JWT signature: %w", err)
+	}
 
-		if err := validator.New().Struct(claims.Custom); err != nil {
-			logging.L.Debug("JWT custom claims not valid")
-			c.AbortWithStatus(http.StatusUnauthorized)
+	key, err := j.getJWK()
+	if err != nil {
+		return c, fmt.Errorf("get JWK from server: %w", err)
+	}
 
-			return
-		}
+	var allClaims struct {
+		jwt.Claims
+		claims.Custom
+	}
+	if err := tok.Claims(key, &allClaims); err != nil {
+		return c, fmt.Errorf("invalid token claims: %w", err)
+	}
 
-		c.Set("name", claims.Name)
-		c.Set("groups", claims.Groups)
+	err = allClaims.Claims.Validate(jwt.Expected{Time: time.Now().UTC()})
+	switch {
+	case errors.Is(err, jwt.ErrExpired):
+		return c, err
+	case err != nil:
+		return c, fmt.Errorf("invalid JWT %w", err)
+	}
 
-		c.Next()
+	if allClaims.Custom.Name == "" {
+		return c, fmt.Errorf("no username in JWT claims")
 	}
+
+	return allClaims.Custom, nil
 }
 
-func proxyMiddleware(proxy *httputil.ReverseProxy, bearerToken string) gin.HandlerFunc {
+func proxyMiddleware(
+	proxy *httputil.ReverseProxy,
+	authn *authenticator,
+	bearerToken string,
+) gin.HandlerFunc {
 	return func(c *gin.Context) {
-		name, ok := c.MustGet("name").(string)
-		if !ok {
-			logging.S.Debug("required field 'name' not found")
-			c.AbortWithStatus(http.StatusUnauthorized)
-
-			return
-		}
-
-		groups, ok := c.MustGet("groups").([]string)
-		if !ok {
-			logging.S.Debug("required field 'groups' not found")
-			c.AbortWithStatus(http.StatusUnauthorized)
-
-			return
-		}
-
-		if name != "" {
-			c.Request.Header.Set("Impersonate-User", name)
-		} else {
-			logging.S.Debug("unable to determine identity")
+		claim, err := authn.Authenticate(c.Request)
+		if err != nil {
+			logging.S.Infof("failed to authenticate request: %v", err)
```

Comment:
Previously we only logged these errors at debug level. I changed it to `Info` because there's no other way to see problems like failing to investigate these problems, and a user shouldn't need to enable debug logs to figure out why they can't connect.